### ISSUE TITLE
Can't get effective breakpoints on x86 application

### ISSUE THREAD

**Issue by Smileynator** [NationalSecurityAgency/ghidra#3641]
I will pose this as a question, because i don't know enough yet to mark this as a bug or not. I am trying to get into this stuff from a programming background.

Going off of the few guides that i found, i am trying to link an older x86 application with the ghidra debugger. (small time unprotected indie app from 2002, nothing cryptic or hard by any definitions)

I can start the application, it breaks at what i assume is program start, i then tell it to resume, and it goes.
What then happens is a few threads start, and not much else. I can still pause the program at will and inspect some memory at that random point. But otherwise i am just dead in the water.

From my understanding i should be able to place breakpoints, but those all stay grey. And while i see like 6 threads, and i can select them, none of them have anything showing in the stack (except 1, that just always says the same static PC)

Interpreter shows modloads, a ton of them (including ntdll.dll, opposed to what debug console seems to claim). And the debug console lists a lot of warnings and errors, which i assume is an issue.

Besides that, when i do pause. And look at memory, any pointers in structures will tell me "i can't show this pointer, it needs to be 8 bytes but the size was 4 bytes" even though if you inspect the type i gave them (mystruct *, Pointer, doesn't seem to matter) they definitely say, 4 bytes in size. So i believe there to be a miscommunication on the pointer sizes the application uses as well. Not sure if that is relevant to this though.
![image](https://user-images.githubusercontent.com/3433068/142192881-70faa3aa-10ca-49fc-ab2d-bdb7524eb7fc.png)

Included is a screenshot of what my debug log looks like after resuming the application from a fresh start of the debugger. Anyone have any clues for me on what i am missing or any guides i could follow to understand this better?
![image](https://user-images.githubusercontent.com/3433068/142192102-f5debfb2-84a2-4998-8ec0-7b3539431739.png)

**Comment by Smileynator** [comment]
Additional basic information to confirm it is a 32bit program etc. I am not sure what else i can tell you about it that is useful.
![image](https://user-images.githubusercontent.com/3433068/142195147-80211fee-1ff3-4a92-8581-03339debc56b.png)

**Comment by d-millar** [comment]
I think we need a little more context here, i.e. how did you connect?  which debugger agent are you using?  are the views populated (not including Object and the Interpreter?  is the view you included the Static or Dynamic listing?  have you imported your game into the Static listing? 

The "module ntdll.dll not found in project" message is telling you you haven't imported that file.  Not necessary, but not a reflection of what has been loaded.  The "readMemory" error messages are probably a result of scrolling through the Dynamic memory while the process is running.  Probably not an issue.

Re breakpoints, if you're setting them in the Static listing, you need to be stopped and to have that program mapped.  Are those both true?

**Comment by Smileynator** [comment]
@d-millar To answer your questions:
I think i was using dbgend locally IN-VM. Though all the others except for the SSH one, worked similarly, and gave similar results. I don't use any other tools besides letting Ghidra boot the application using one of these 4 options.
Views are populated, but as described, but very minimalistic (i see a few stackpointers, but nothing sane, i see threads running and closed, etc.) If you want me to check specifics at any point, tell me which are important things to look at.
The view i included was the Dynamic listing on the upper half, and static listing below the dividing line.
I was not even aware one could load one without the other, so yes, they both are the same file, and both loaded with data from that file. But paused application at the time of screenshot

As for breakpoints. When i set them or where, doesn't seem to affect what is happening. But yes, i was stopped at the time of adding. I tried all different combinations i could think of, not seeing any difference.
But to be clear, paused program, using K to toggle breakpoint, in the static view, on a CALL asm line, automatically recognizes as SW_EXECUTE. But stays with a grey dot.
As for "having the program mapped" i am not sure what you imply. But the breakpoint was placed in a function that was recognized by ghidra as such. But did not trigger as far as i could tell.

I then used F5 to resume the program (as it was auto breaked at start i believe?) and it boots. And as far as i can read assembly this method HAS to get called (it initializes a lot of globals used for file access that the rest of the code uses) but nothing happens, the game loads into main menu, and i am left with this (note i did not touch a thing after F5)
![image](https://user-images.githubusercontent.com/3433068/142764026-af68a8a0-6b53-468a-a027-41b40406cc09.png)
Nothing happens here even if i wait forever. If i pause i randomly end up in one of the threads, and by certain data things at certain memory locations i can quickly tell if that is the main process or not. (often is, but not always)

If i then close the application (as in the actual running application, not killing the debugger) it doesn't even seem to realize that the process died at all. Making me further suspect that i am not properly connected. (Then again, i have little experience with Ghidra so far either way, so maybe it's just how things work around here)

**Comment by d-millar** [comment]
@Smileynator The screenshot was very helpful - thanks for posting that.  I have a couple of guesses.  With luck, we can figure out which one is the problem....

A couple of comments first:  in your latest screenshot, the Dynamic listing is showing stale memory indicated by the fact that its' grey.  Hitting fresh in the Objects view, "should" fix this.  This results form a failure (on our part) to aggressively update the memory.  We were shooting for rapid updates, but there's some corner case I'm not handling well.  Apologies.  It's on the list of things to fix.  

Obviously, this isn't the crux of the problem for you because your earlier screenshot shows current memory.  What I can't tell from the earlier screenshot is whether the Dynamic & Static listings are tracking together while you're executing in the Static listing's program space.  This matters because it would tell me whether the Static program is mapped correctly into the running target's address space.  You should be able to tell this from the Modules view.

If the Static view is not mapped correctly, setting breakpoints there will not do much.  Specifically, if that's true, you are in essence setting "future" breakpoints - breakpoints that are saved and activated only after the program is mapped.  You can verify the current set of active breakpoints in the Objects view or in the Interpreter window using "bl".  Have you tried verifying that your breakpoint hits in straight-up, Ghidra-free windbg?  If so, have you tried setting the same breakpoint from either the Interpreter command line or the Objects view?  As I think I mentioned, those two are the most likely to reflect the state of the debugger.  The other windows depend on the state of the recording and, in the case of breakpoints, the state of the static-dynamic mapping.

If the Static/Dynamic views are not stepping together, the most likely cause is that the name of the program is not a clear match for the name of the executable in the Module list (or the Module view is empty or lacks the target program completely).  If so, you can either (a) rename your executable in the Static listing, or (b) map the program by hand.  To map it by hand, go to the Module view, select the current executable, right-click, and choose  "Map Module to blahblahblah".  (There also may be an error message in the Console window with a button that will fix the problem.)

Let us know if any of this fails or doesn't make sense.  We definitely want to get this working for you!

Best,
D

**Comment by Smileynator** [comment]
@d-millar Thanks for the help so far. I guess it must be frustratingly hard to make this work, and then a relative starter is breaking it because he has no clue how delicate it is.

Ok, I have started in the programming like before. Tracking in the dynamic window is fixed at one position with all ?? 00h like you have seen before. The static window never moves from it's original spot (in this case since i didn't touch it. The program entry point 0x400000)

Switching the dynamic window to track PC instead of SP, makes no difference besides that it jumped to a really high address range, where it proceeded to not move to any other address and show ?? 00h all over.

Pressing the refresh button on Objects seems to make no visual difference.

You then say something about the modules view. The modules view shows me all of the loaded DLL's it seems. And none of them match the name of the EXE. But before i "resume" the program after the initial break, the only module there is that first one Image0000-.
![image](https://user-images.githubusercontent.com/3433068/142773317-f65b61cb-96bc-464b-b2ed-b8f455790a96.png)

You then mention some stuff i have not a lot of knowledge about, the interpreter. I tried typing "bl" without quotes straight into the window (seems to pick thread 1 by default, which later closed to make place for thread 2 as default) and it gave me no response. Suffice to say i have never ran a debugger outside of visual studio's default on my own written code in debug mode, and have no clue what to expect here. (If you have something that is a minimum must read on "how to debug" in this sense, do let me know)
![image](https://user-images.githubusercontent.com/3433068/142773424-ac3bf30c-8ee2-4e6c-b571-17f95d85a578.png)

Lastly, i tried to do what you told me by remapping it to what i suspected to be the actual EXE (as in, all those DLL's don't match my exe name in the slightest, nor do any of them look like they could be)
![image](https://user-images.githubusercontent.com/3433068/142773503-265deeeb-b96c-4cd4-a601-fc055e14f328.png)
Pressing Ok in the above, again made no change to any of the other open pages. Tracking doesn't seem to be happening.

When i get back from dinner, i will try to rename the EXE in the static listing, but i doubt it will yield different results if they should amount to the same.

**Comment by d-millar** [comment]
Yes, the debugger is definitely a work in progress.  We do appreciate your trying it out.  Folks who are relatviely new to debugging are our best test drivers - they find and help us fix more errors than anyone else.  So, big thanks!

You are defintely making progress here, in any case.  Image00000 is your guy, so mapping him to your main program was the right thing to do.  To test out the mapping, you can try a few things.  First, make sure the Dynamic Listing is set to track the PC.  Second, map the main program the way you did (maybe, make sure you select it before you hit OK - but I think you did the right thing).  Hitting G (go) in the Static listing and entering an address, say, of where you want to break, should move both the Static and Dynamic listings to that address.  (I might be wrong about that.)  Also, if, after hitting the initial breakpoint, you hit step, instead of using a breakpoint, both windows should track.  If not, we need to do a little more work.

If that does work, also try setting the breakpoint (using K) in either the Static or Dynamic listing (after the program is started) and then try bl in the Interpreter.  If it's not there, we also have more work to do.  If it is there but isn't hitting, try clicking around in your target.  Sometimes the thread that trips the break point doesn't fire until the target gets focus.

**Comment by Smileynator** [comment]
Ok, interesting. Seemingly i did do the right thing then with mapping (i did it again with selection, it has a short few tiny popups that almost immediately go away, and that is all it does)

Using G or navigating to anywhere, does indeed make both static and dynamic listing track, though dynamic is still all ?? 00h about the whole thing.

On the side i noticed something weird, probably not relevant. When i used K to set a breakpoint in the static listing now. And hit OK to the dialog. It doesn't go away. It adds the breakpoint (still grey) to the breakpoints window. But the popup doesn't close for some reason.

BL still has the same behavior (no response after hitting return)

Taking it all the way from the top (mapping BEFORE resuming) i think it now works as intended. I can walk through the code once the breakpoint hits, it tracks properly, and i think this resolves the issue for getting breakpoints working. Great!
(This also fixes tracking, BL returns 1 breakpoint, etc.)

Now that leaves me with 2 burning questions:
1 - Is there any way to tell Ghidra "that Image00000 is the thing to auto attach to"?
2 - Is there any specific reason why even my valid breakpoints, do not last between sessions? (they turn grey again) Expected behavior for me, would be "this breakpoint is associated with this mapping/exe, when the mapping/exe becomes valid, re-apply all "future breakpoints" as you named them. Especially with the eye on frequent restarting of the debugger/exe to process specifically the start-up sequence of the software in mind.
3 - As i can now properly view the structure i was originally looking at, as it is being edited, i am still somewhat blindsighted by that HModule pointer requiring 8 Bytes, even though that is a thing auto generated by Ghidra, and should be 4Bytes by all accounts.

By any other accounts, besides being finicky because it is a rather new addition, it looks like it doing it's thing right now. So i can experiment away with runtime information. Many thanks for the help once more!

**Comment by d-millar** [comment]
Awesome - very glad you got it (more or less!) working.  Thanks too for the pointer on mapping BEFORE launching. 

With respect to your questions:

(1) Right now, the only way that I know of doing this is to rename the program.  We could probably come up with a more permanent mapping.  I'm just not sure where exactly to store it.  Will put in a ticket and discuss with the other debugger folks.

(2) I agree.  That's exactly how they should behave.  Not clear to me at the moment whether they aren't because of (1) or because the Windows debugger implementation is lacking.  (This does work correctly for GDB/Linux targets, but entirely possible I overlooked some pieces in the Windows equivalent.)  Either way, that's definitely a new ticket.  Will let you know when I know a bit more.

(3) So, the Ghidra data structure libraries are created by parsing the OS-specific header files and matched to the platform you're running on.  In this case, HModule (if I remember correctly) is defined to be an HINSTANCE in WinDef.h (?), which in turn is defined to be HANDLE which is treated as a pointer.  (Not sure why this is the case actually - it's not really a pointer, but....)  In any case, on a 64-bit machine, am pretty sure HANDLEs are 64-bit.  You're running in a WoW environment, and understandably would like the structures to reflect that.  Unfortunately, the structure libraries have no knowledge of the address space where you're applying them (I think) and therefore are defaulting to 8-byte.  Will ask around tomorrow - see if there's a workaround for this.  Or whether I'm actually completely wrong.  :) 

More soon - best, D

**Comment by Smileynator** [comment]
Ok, last comments, and then i will leave you be on this specific ticket :)

(1) I assume this can just be some Ghidra project metadata that can be stored (known mapping names) especially if you are linking up more than 1 file (which the thing seems to support from a glance) this could be helpful if you can only resort to manual linking. (renaming my EXE to image000'00004000 surely will not go over well with windows, unless that is an internal name to ghidra?)

(2) Let me be clear in that they ARE around (in the upper half of the breakpoints window) and i can mass enable them (at any point after mapping anyway) and they behave as expected. It's just an annoying manual stop that it sort of acts like "auto disable" every time i stop debugging. Like in visual studio with your own programs, you kind of expect them to persist.
Tag me on that new ticket, i'd love to test any new versions (then again, i can't compile ghidra myself at the moment, so i might not be of great help)

(3) I suspected something like this was going on. I have no clue on the internal workings here. But it would be great if we could somehow tell it (or technically, it knows already!) that this is a x86 environment, and to expect 4byte pointers while inside it's data structures. I am not sure how much it affects actual use though, besides the warning texts. As long as it still functions as long as you ignore the warning, that would be great, but i have a feeling it's not that easy. Keep me up to date on this one as well, i am sure this will apply to many an old game fanatic trying to do some tampering.

**Comment by Wall-AF** [comment]
> (3) I suspected something like this was going on. I have no clue on the internal workings here. But it would be great if we could somehow tell it (or technically, it knows already!) that this is a x86 environment, and to expect 4byte pointers while inside it's data structures. I am not sure how much it affects actual use though, besides the warning texts. As long as it still functions as long as you ignore the warning, that would be great, but i have a feeling it's not that easy. Keep me up to date on this one as well, i am sure this will apply to many an old game fanatic trying to do some tampering.

@Smileynator Quick question: are you saying the _normal_ pointers are 8 bytes and _pointers within structures_ are 4 bytes?  If so, do those 4 byte pointers have a _hidden_ addition 4 bytes (like in segmented architectures)?

**Comment by Smileynator** [comment]
@Wall-AF I have no idea what the intended way would be for this to work.
But how i assume it works, is that x86 applications are ran in some sort of sandbox, in order to deal with the whole "32bits thing" including pointer sizes and probably some other shenanigance.
When an application is identified as 32bit, it would make sense for Ghidra to assume that pointers in said application can not magically be 64bit long just because the host machine is.
So either it translates and shows us the "value that is given to the 32bit emulation" or it shows us the 64bit pointer value, but doesn't warn us about the non fitting size (rather it should say "4bytes assigned, 8bytes emulated" or something along those lines)
Something, that in the context, would make sense and reflects what is actually happening in the targeted application.
Again, i don't know the technical side of it, i never really figured out how 32bit compatibility works under the hood. But i think you grasp what i am trying to convey.

**Comment by Smileynator** [comment]
Summary:
1) Is a new ticket i specified as #3675 with the relevant details.
2) Is a ticket internally that will need some work, nothing we can do about that now on the windows version.
3) Awaiting response about pointers in a WoW64 environment. 
Notable addition: during a debugging session (as in, actively debugging) it becomes impossible for me to define pointers as datatypes, because even in the structure editor they become the assumed 64bit pointer. Thus char* == 8 bytes, not the expected x86 4 bytes. Even though it does still list as 4 bytes.
![image](https://user-images.githubusercontent.com/3433068/143624476-21452338-f851-4516-91d2-d85d3b76e1a1.png)
![image](https://user-images.githubusercontent.com/3433068/143624504-1014bb1a-0eba-4539-b997-5591e1c8f7e0.png)

**Comment by Smileynator** [comment]
And after rummaging through some documentation.
You can apparently use "char*32" it will then just put it in, it will look like a normal char*, but will be the 32bit size.
It will also not show the warning message about pointers in the dynamic window. But it is unintuitive, more typing work, and a bit silly that typing it differently, yields different results, but similar visual representation. And even more so that THEN all of sudden the rest of ghidra is totally fine with this pointer not being 64 bit. Even though all other pointers i made before were just written as "char*" and accepted as 32bit by default.
Not sure what to make of this as an end user. I hope it says more towards the devs.
